src/java/org/apache/cassandra/index/sai/disk/v1/SegmentMetadata.java

```diff
@@ -0,0 +1,362 @@
+    {
+        assert numRows < Integer.MAX_VALUE;
+        Objects.requireNonNull(minKey);
+        Objects.requireNonNull(maxKey);
+        Objects.requireNonNull(minTerm);
+        Objects.requireNonNull(maxTerm);
+
+        this.segmentRowIdOffset = segmentRowIdOffset;
+        this.minSSTableRowId = minSSTableRowId;
+        this.maxSSTableRowId = maxSSTableRowId;
+        this.numRows = numRows;
+        this.minKey = minKey;
+        this.maxKey = maxKey;
+        this.minTerm = minTerm;
+        this.maxTerm = maxTerm;
+        this.componentMetadatas = componentMetadatas;
+    }
+
+    private SegmentMetadata(IndexInput input, PrimaryKeyFactory primaryKeyFactory) throws IOException
+    {
+        this.segmentRowIdOffset = input.readLong();
+        this.numRows = input.readLong();
+        this.minSSTableRowId = input.readLong();
+        this.maxSSTableRowId = input.readLong();
+        this.minKey = primaryKeyFactory.createPartitionKeyOnly(DatabaseDescriptor.getPartitioner().decorateKey(readBytes(input)));
+        this.maxKey = primaryKeyFactory.createPartitionKeyOnly(DatabaseDescriptor.getPartitioner().decorateKey(readBytes(input)));
+        this.minTerm = readBytes(input);
+        this.maxTerm = readBytes(input);
+        this.componentMetadatas = new ComponentMetadataMap(input);
+    }
+
+    @SuppressWarnings({"resource", "RedundantSuppression"})
+    public static List<SegmentMetadata> load(MetadataSource source, PrimaryKeyFactory primaryKeyFactory) throws IOException
+    {
+        IndexInput input = source.get(NAME);
+
+        int segmentCount = input.readVInt();
+
+        List<SegmentMetadata> segmentMetadata = new ArrayList<>(segmentCount);
+
+        for (int i = 0; i < segmentCount; i++)
+        {
+            segmentMetadata.add(new SegmentMetadata(input, primaryKeyFactory));
+        }
+
+        return segmentMetadata;
+    }
+
+    /**
+     * Writes disk metadata for the given segment list.
+     */
+    public static void write(MetadataWriter writer, List<SegmentMetadata> segments) throws IOException
+    {
+        try (IndexOutput output = writer.builder(NAME))
+        {
+            output.writeVInt(segments.size());
+
+            for (SegmentMetadata metadata : segments)
+            {
+                output.writeLong(metadata.segmentRowIdOffset);
+                output.writeLong(metadata.numRows);
+                output.writeLong(metadata.minSSTableRowId);
+                output.writeLong(metadata.maxSSTableRowId);
+
+                Stream.of(metadata.minKey.partitionKey().getKey(),
+                          metadata.maxKey.partitionKey().getKey(),
+                          metadata.minTerm, metadata.maxTerm).forEach(bb -> writeBytes(bb, output));
+
+                metadata.componentMetadatas.write(output);
+            }
+        }
+    }
+
+    @Override
+    public int compareTo(SegmentMetadata other)
+    {
+        return Long.compare(this.segmentRowIdOffset, other.segmentRowIdOffset);
```

Comment:
I'm guessing we only compare within the same column index and SSTable?

Reply:
I've had a look at it's usage and `SegmentMetadata` is never compared so I've removed the `compareTo` method.